src/main/java/server/ChatServer.java

```diff
@@ -8,118 +8,78 @@
+        t.start();
+    }
 
-    public static void main(String[] args) throws IOException {
-        // list of approved users
-        UserLogin userlogin = new UserLogin();
-        List<String> userList = userlogin.listOfUsers();
-
-        // server is listening on port 8000
-        ServerSocket ss = new ServerSocket(8000);
-        System.out.println("Server started");
-        Socket s;
-
-        // running infinite loop for getting
-        // client request
-        while (true) {
-            // Accept the incoming request
-            s = ss.accept();
-
-            System.out.println("New client request received : " + s);
-
-            // obtain input and output streams
-            DataInputStream dis = new DataInputStream(s.getInputStream());
-            PrintWriter dos = new PrintWriter(s.getOutputStream(),true);
-            StringTokenizer st = new StringTokenizer(dis.readLine(), "#");
-            String cmd = st.nextToken();
-
-            try {
-
-
-                if (cmd.equals("CONNECT")) {
-                    String username = st.nextToken();
-                    int i = 0;
-
-
-                    for (String users : userList) {
-
-
-                        if (username.equals(users)) {
-
-                            System.out.println("Creating a new handler for this client...");
-
-                            // Create a new handler object for handling this request.
-                            ClientHandler mtch = new ClientHandler(s, username, dis, dos);
-
-                            // Create a new Thread with this object.
-                            Thread t = new Thread(mtch);
-
-                            System.out.println("Adding this client to active client list");
-
-                            // add this client to active clients list
-                            ar.add(mtch);
-
-                            // start the thread.
-                            t.start();
-
-                            onlineMessage();
-
-                            break;
-
-                        }
-                        i++;
-
-                        if (i == userList.size()) {
-                            dos.println("CLOSE#2");
-                            s.close();
-                            break;
-                        }
-
-
-                    }
+    /**
+     * sendMessage attempts to send a message to the given destination.
+     */
+    public static void sendMessage(ClientHandler source, String destination, String message) {
+        StringBuilder builder = new StringBuilder();
+        builder.append("MESSAGE#");
+        if (destination == "*") {
+            builder.append("*");
+        } else {
+            builder.append(source.getName());
+        }
 
+        builder.append("#");
+        builder.append(message);
 
-                }
 
+        String msg = builder.toString();
+        for (ClientHandler mc : ChatServer.ar) {
+            if (!mc.nameMatches(pattern)) {
```

Comment:
We move the matching logic to `ClientHandler` to avoid having to ask `ClientHandler` what its name is. Getters are a crutch in OOP and should be avoided unless necessary as they break encapsulation. It's better to tell the object your intent and let it work out what to do next.